parallel/runner.go

```diff
@@ -158,26 +143,112 @@ func (r *runner) Errors() map[int]error {
 func (r *runner) DoneWhenAllIdle(idleThresholdSeconds int) error {
 	for {
 		time.Sleep(time.Duration(idleThresholdSeconds) * time.Second)
-		for i := 0; i < r.maxParallel; i++ {
-			if !r.idle[i] {
-				break
+		allIdle := true
+		var e error
+		r.idle.Range(func(key, value interface{}) bool {
```

Comment:
Let's explain what we're iterating over here and what's the goal of this iteration. Let's also explain the boolean returned value.